fetching/tarball-fetcher/src/remoteTarballFetcher.ts

```diff
@@ -183,35 +186,41 @@ export function createDownloader (
             reject(err)
             return
           }
-          const data: Buffer = Buffer.from(new ArrayBuffer(downloaded))
+          const data: Buffer = Buffer.from(new SharedArrayBuffer(downloaded))
           let offset: number = 0
           for (const chunk of chunks) {
             chunk.copy(data, offset)
             offset += chunk.length
           }
+          const localWorker = await pool.checkoutWorkerAsync(true)
 
           try {
             if (opts.integrity) {
-              try {
-                ssri.checkData(data, opts.integrity, { error: true })
-              } catch (err: any) { // eslint-disable-line
-                throw new TarballIntegrityError({
-                  algorithm: err.algorithm,
-                  expected: err.expected,
-                  found: err.found,
-                  sri: err.sri,
-                  url,
-                })
+              const [, algo, integrityHash] = opts.integrity.match(INTEGRITY_REGEX)!
+              // Compensate for the possibility of non-uniform Base64 padding
+              const normalizedRemoteHash: string = Buffer.from(integrityHash, 'base64').toString('hex')
+
+              const calculatedHash: string = crypto.createHash(algo).update(data).digest('hex')
+              if (calculatedHash !== normalizedRemoteHash) {
+                throw new Error(
+                  `integrity validation failed:\nintegrity: ${opts.integrity}\nexpected: ${normalizedRemoteHash}\nreceived: ${calculatedHash}`
+                )
               }
```

Comment:
I recommend moving the hash computation to the worker. It can still send a rejection message to the main thread that can then get surfaced as an error.

Reply:
Done